src/solver.py

```diff
@@ -77,159 +87,41 @@ class SolverConfig:
 
 
-class GlobalHandler:
-    def __init__(self,
-                 global_vars: Set[DerivedTypeVariable],
-                 callgraph: networkx.DiGraph,
-                 scc_dag: networkx.DiGraph,
-                 fake_root: DerivedTypeVariable) -> None:
-        self.scc_dag = scc_dag
-        self.callgraph = callgraph
-        self.global_vars = global_vars
-        self.fake_root = fake_root
-
-    def pre_scc(self, scc_node: Any):
-        raise NotImplementedError("Child class must implement")
-
-    def post_scc(self,
-                 scc_node: Any,
-                 sketches_map: Dict[DerivedTypeVariable, "Sketches"]) -> None:
-        raise NotImplementedError("Child class must implement")
-
-    def copy_globals(self,
-                     current_sketch: "Sketches",
-                     nodes: Set[DerivedTypeVariable],
-                     sketches_map: Dict[DerivedTypeVariable, "Sketches"]) -> None:
-        raise NotImplementedError("Child class must implement")
-
-    def finalize(self,
-                 solver: "Solver",
-                 sketches_map: Dict[DerivedTypeVariable, "Sketches"]) -> None:
-        raise NotImplementedError("Child class must implement")
-
-
-class PreciseGlobalHandler(GlobalHandler):
+class EquivRelation:
     """
-    Handle globals precisely, by bubbling sketches up the callgraph, accumulating more and more
-    globals (and sketch nodes) as it goes. Once it reaches the roots of the callgraph, we have
-    all the globals and these are the final sketches.
+    This class represents an equivalence relation
+    that can be computed incrementally
     """
-    def __init__(self,
-                 global_vars: Set[DerivedTypeVariable],
-                 callgraph: networkx.DiGraph,
-                 scc_dag: networkx.DiGraph,
-                 fake_root: DerivedTypeVariable) -> None:
-        super(PreciseGlobalHandler, self).__init__(global_vars, callgraph, scc_dag, fake_root)
-        # SCCs that have not yet had their sketches cleansed of globals
-        # TODO get proper type annotations
-        self.not_cleaned_up: List[Tuple[Any, Set[Any]]] = []
-
-
-    def cleanse_globals(self, sketches: "Sketches"):
-        """
-        Delete all the global nodes from a particular sketch graph.
-        """
-        for dtv, node in list(sketches.lookup.items()):
-            if dtv.base_var in self.global_vars:
-                if node in sketches.sketches.nodes:
-                    sketches.sketches.remove_node(node)
-                del sketches.lookup[dtv]
 
+    def __init__(self, elems: Set[DerivedTypeVariable]) -> None:
+        self._equiv_repr = {elem: frozenset((elem,)) for elem in elems}
 
-    def pre_scc(self, scc_node: Any) -> None:
-        """
-        Sets up a list of all SCCs that still have callers to be processing. See post_scc for
-        what happens when all callers have been processed.
+    def make_equiv(
+        self,
+        x: FrozenSet[DerivedTypeVariable],
```

Comment:
We use it to partition the nodes in the graph used for `inferShapes`, that partition is the used to compute the quotient graph.

Reply:
Ok. No action needed now. I mentioned it because sets that only do union and find can operate in almost-constant time (inverse Ackermann). If this turns out to be significant for performance, we can tweak this. For now, `frozenset` is correct.

Reply:
Makes sense, I'll add a TODO so we don't forget ;)